cmd/antrea-agent/options_windows.go

```diff
@@ -61,3 +66,119 @@ func (o *Options) checkUnsupportedFeatures() error {
 	}
 	return nil
 }
+
+func registerService() error {
+	p, err := getServicePath()
+	if err != nil {
+		return err
+	}
+	m, err := mgr.Connect()
+	if err != nil {
+		return err
+	}
+	defer m.Disconnect()
+
+	c := mgr.Config{
+		ServiceType:  windows.SERVICE_WIN32_OWN_PROCESS,
+		StartType:    mgr.StartAutomatic,
+		ErrorControl: mgr.ErrorNormal,
+		DisplayName:  "Containerd",
+		Description:  "Container runtime",
+	}
+
+	// Configure the service to launch with the arguments that were just passed.
+	args := []string{"--run-service"}
+	for _, a := range os.Args[1:] {
+		if a != "--register-service" && a != "--unregister-service" {
+			args = append(args, a)
+		}
+	}
+
+	s, err := m.CreateService(serviceNameFlag, p, c, args...)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	// See http://stackoverflow.com/questions/35151052/how-do-i-configure-failure-actions-of-a-windows-service-written-in-go
+	const (
+		scActionNone    = 0
+		scActionRestart = 1
+
+		serviceConfigFailureActions = 2
+	)
+
+	type serviceFailureActions struct {
+		ResetPeriod  uint32
+		RebootMsg    *uint16
+		Command      *uint16
+		ActionsCount uint32
+		Actions      uintptr
+	}
+
+	type scAction struct {
+		Type  uint32
+		Delay uint32
+	}
+	t := []scAction{
+		{Type: scActionRestart, Delay: uint32(15 * time.Second / time.Millisecond)},
+		{Type: scActionRestart, Delay: uint32(15 * time.Second / time.Millisecond)},
+		{Type: scActionNone},
+	}
+	lpInfo := serviceFailureActions{ResetPeriod: uint32(24 * time.Hour / time.Second), ActionsCount: uint32(3), Actions: uintptr(unsafe.Pointer(&t[0]))}
+	err = windows.ChangeServiceConfig2(s.Handle, serviceConfigFailureActions, (*byte)(unsafe.Pointer(&lpInfo)))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getServicePath() (string, error) {
+	p, err := exec.LookPath(os.Args[0])
+	if err != nil {
+		return "", err
+	}
+	return filepath.Abs(p)
+}
+
+func unregisterService() error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return err
+	}
+	defer m.Disconnect()
+
+	s, err := m.OpenService(serviceNameFlag)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	err = s.Delete()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// registerUnregisterService is an entrypoint early in the daemon startup
+// to handle (un-)registering against Windows Service Control Manager (SCM).
+// It returns an indication to stop on successful SCM operation, and an error.
+func registerUnregisterService(root string) (bool, error) {
```

Comment:
I was wondering where would we invoke the registerUnregisterService, since it would self register would we be doing this after the check if the Node is windows/linux ? If yes I could not find that code in the codebase
CC @jayunit100

Reply:
@perithompson

Reply:
I think typically you can leverage the go build tag to check the OS (so don't need to check via a `if` condition explicitly), i.e.:

- Name it more generically, like `initForOS` (from https://github.com/kubernetes/kubernetes/blob/6af7eb6d494d8f742e18a3b965053e3a518bf6d7/cmd/kubelet/app/init_windows.go#L47)
- Have a dummy linux version in *_linux.go
- Call `initForOS` in OS generic code, i.e. `cmd/antrea-agent/agent.go`